devtools/server/actors/replay/connection.js

```diff
@@ -533,6 +527,117 @@ class Recording extends EventEmitter {
   }
 }
 
+function checkShouldValidateUrl() {
+  if (gShouldValidateUrl !== null) {
+    return Promise.resolve(gShouldValidateUrl)
+  } else {
+    return queryAPIServer(`
+      query GetOrgs {
+        viewer {
+          workspaces {
+            edges {
+              node {
+                isOrganization
+                settings {
+                  features
+                }
+              }
+            }
+          }
+        }
+      }
+    `).then(resp => {
+      if (resp.errors) {
+        throw new Error("Unexpected error checking Replay user permissions");
+      }
+
+      const workspaces = resp.data.viewer?.workspaces.edges;
+      gShouldValidateUrl = !workspaces ? false : workspaces.some(w => {
+        if (w.node.isOrganization) {
+          const {allowList, blockList} = w.node.settings?.features?.recording || {};
+
+          return (Array.isArray(allowList) && allowList.length > 0) || (Array.isArray(blockList) && blockList.length > 0);
+        }
+      });
+
+      return gShouldValidateUrl;
+    });
+  }
+}
+
+async function canRecordUrl(url) {
+  try {
+    const shouldValidate = await checkShouldValidateUrl();
+    if (!shouldValidate) return true;
+
+    const resp = await queryAPIServer(`
+      query CanRecord ($url: String!) {
+        viewer {
+          canRecordUrl(url: $url)
+        }
+      }
+    `, {
+      url
+    });
+
+    if (resp.errors) {
+      throw new Error(resp.errors[0].message);
+    }
+
+    return resp.data.viewer.canRecordUrl;
+  } catch (e) {
+    // Fallback to allowing recordings if the backend errors but log to telemetry
+    console.error(e);
+    pingTelemetry("recording", "can-record-failed", { why: e.message || "", url });
+
+    return true;
+  }
+}
+
+function getLocationListener(key) {
+  return {
+    key,
+    QueryInterface: ChromeUtils.generateQI([
+      "nsIWebProgressListener",
+      "nsISupportsWeakReference",
+    ]),
+    onLocationChange(aWebProgress, _, aLocation) {
+      if(!aWebProgress.isTopLevel) return;
+
+      // Always allow blank and new tab
+      if (aLocation.displaySpec === "about:blank" || aLocation.displaySpec === "https://app.replay.io/browser/new-tab") {
+        return;
+      }
+
+      canRecordUrl(aLocation.displaySpec).then((canRecord) => {
+        if (canRecord) return;
+
+        const browser = getRecordingBrowser(this.key);
```

Comment:
What if anything should happen if I navigate the tab to a disallowed location but then press stop while we're still waiting for the response to this? Seems like we'd allow the recording through.

Reply:
I think we'd allow it at this point